lib/streamlit/runtime/caching/cache_utils.py

```diff
@@ -377,18 +387,15 @@ def get_or_create_cached_value():
                 messages = cached_func.message_call_stack._most_recent_messages
                 try:
                     cache.write_result(value_key, return_value, messages)
-                except TypeError:
-                    if type_util.is_type(
-                        return_value, "snowflake.snowpark.dataframe.DataFrame"
-                    ):
-
-                        class UnevaluatedDataFrameError(StreamlitAPIException):
-                            pass
-
+                except (TypeError, RuntimeError, CacheError):
```

Comment:
This seems overly broad - is it actually possible for any of these errors to be raised from an uncollected dataframe? (My assumption is that it's really just a `CacheError` that could be raised from a pickling failure?)

Reply:
So the reason I've introduced 3 types of exceptions here are:
- Snowpark triggers TypeError
- Spark triggers RunTime error in the Testing environment, and CacheError when run on normal build
Is that ok?

Reply:
I'd love to narrow this down - what are the operations that cause the `TypeError` and `RuntimeError` errors?

in `get_or_create_cached_value`, we’re catching exceptions thrown by `cache.write_result(value_key, return_value, messages)`. The st.experimental_memo implementation of `cache.write_result` uses the pickle utility to get the bytes to write to the cache:

```python
try:
    pickled_entry = pickle.dumps(multi_cache_results)
except pickle.PicklingError as exc:
    raise CacheError(f"Failed to pickle {key}") from exc
```

If we get `TypeError` and `RuntimeError` thrown from write_result, then it seems like it must be happening elsewhere in the function?